### PR TITLE
BFD-2405: Establish SLIs for Pipeline SLOs

### PR DESCRIPTION
**Pull request by malessi**
**JIRA Ticket:**
[BFD-2405](https://jira.cms.gov/browse/BFD-2405)

**User Story or Bug Summary:**

As a BFD Engineer, I want to have data available describing the behavior of the BFD Pipeline such
that I can derive SLIs that can be used for our agreed-upon BFD Pipeline SLOs

---

### What Does This PR Do?

This PR adds a new submodule to the existing BFD Pipeline "Terraservice" called `bfd_pipeline_slis`.
This new module encapsulates the Terraform IaC and Python source code for a new per-environment
Lambda: `bfd-ENV-update-pipeline-slis`. This Lambda is invoked whenever a new file is uploaded to
the environment's corresponding pipeline/ETL S3 Bucket in either the `Done` or `Incoming` "folders".
The Lambda then puts various metrics up to CloudWatch Metrics based upon the timestamp of the
incoming event notification. These metrics allow for the calculation of various SLIs, such as the
time it takes to load a certain RIF file type, the time it takes to fully load incoming data, the
time at which data was first made available and when it was finally loaded, and more.

The metrics that are put to CloudWatch Metrics, and their dimensions, are as follows:

| Metric | Dimensions (Aggregations) | Data Type | Description
| --- | --- | --- | --- |
| `time/data-available` | - None<br>- `data_type`<br>- `group_timestamp`<br>- `data_type`, `group_timestamp` | UTC Timestamp in seconds | Indicates when data was first made available, in `Incoming`, to the pipeline |
| `time/data-loaded` | - None<br>- `data_type`<br>- `group_timestamp`<br>- `data_type`, `group_timestamp` | UTC Timestamp in seconds | Indicates when data was loaded by the pipeline |
| `time/data-first-available` | - None<br>- `group_timestamp` | UTC Timestamp in seconds | Indicates the time of the first data file, in its group, being made available to the pipeline |
| `time/data-fully-loaded` | - None<br>- `group_timestamp` | UTC Timestamp in seconds | Indicates the time when the final data file, in its group, was loaded by the pipeline |
| `time-delta/data-load-time` | - None<br>- `data_type`<br>- `group_timestamp`<br>- `data_type`, `group_timestamp` | A time delta in seconds | Indicates the amount of time, in seconds, that the pipeline took to load a particular RIF file in a particular group |

With the above metrics, we should have the data required to satisfy the Pipeline's SLOs and provide
useful alerts when the Pipeline is out of SLO.

The changes within this PR have been validated to be correct through manual testing using synthetic
S3 Event Notifications, and through applying the Terraform to the `test` environment and testing the
corresponding Lambda. Note that the changes to the `test` environment have since been undone.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

- Verify all PR security questions and checklists have been completed and addressed.
- Are additional metrics required to satisfy our Pipeline SLOs?
- Does the implementation make sense? Is it too large?

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

- If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

  - Does this PR add any new software dependencies?
    - [ ] Yes
    - [x] No
  - Does this PR modify or invalidate any of our security controls?
    - [ ] Yes
    - [x] No
  - Does this PR store or transmit data that was not stored or transmitted before?
    - [ ] Yes
    - [x] No

- If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
  - Do you think this PR requires additional review of its security implications for other reasons?
    - [ ] Yes
    - [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

- N/A

### Submitter Checklist

I have gone through and verified that...:

- [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
- [x] This PR is reasonably limited in scope, to help ensure that:
  1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
  2. There isn't too much of a burden on reviewers.
  3. Any problems it causes have a small "blast radius".
  4. It'll be easier to rollback if that becomes necessary.
- [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
- [x] The data dictionary has been updated with any field mapping changes, if any were made.
- [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
- [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
- [x] Reviews are requested from both:
  - At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
  - Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
- [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
  - Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+            )
+            backoff_retry(
+                func=put_timestamp_metrics, ignored_exceptions=COMMON_UNRECOVERABLE_EXCEPTIONS
+            )
+            print(f'Successfully put metrics to "{METRICS_NAMESPACE}/{timestamp_metric_name}"')
+        except Exception as exc:
+            print(
+                "An unrecoverable error occurred when trying to call PutMetricData for metric"
+                f" {METRICS_NAMESPACE}/{timestamp_metric_name}: {exc}"
+            )
```
Comment (lsmitchell) [inline ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py:388]: If this is unrecoverable, do we want to return here? (We do in some other unrecoverable scenarios below)

File: ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+                )
+                backoff_retry(
+                    func=put_data_first_available_metrics,
+                    ignored_exceptions=COMMON_UNRECOVERABLE_EXCEPTIONS,
+                )
+                print("Data put successfully")
+            except Exception as exc:
+                print(
+                    f"An unrecoverable error occurred when trying to call PutMetricData; err: {exc}"
+                )
```
Comment (lsmitchell) [inline ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py:474]: Another case where we have an unrecoverable error but don't return; is it that we can continue on the puts failing but not the gets?

**Review by lsmitchell — COMMENTED** [review]
Thanks for going over this in the meeting the other day, made it easier to understand contextually. Just a few minor things.

File: ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+    identified by their namespace, name, and dimensions"""
+
+    metric_namespace: str
+    metric_name: str
+    dimensions: dict[str, str] = field(default_factory=dict)
+
+
+@dataclass
+class MetricDataResult:
+    """Dataclass representing the result of a successful GetMetridData operation"""
```
[the remark this answers is not in the document]
Comment (malessi) [reply ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py]: Thanks for catching this typo!

File: ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py (hunk 1 above)
Comment (malessi) [reply ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py:388]: Good catch, I had forgotten to include the `return` within the catch block. I've added the `return` back.

File: ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py (hunk 2 above)
Comment (malessi) [reply ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py:475]: Same deal as above: looks like I forgot to include a `return` here. Thanks for catching this!

File: ops/terraform/services/pipeline/modules/bfd_pipeline_slis/iam.tf [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+resource "aws_iam_policy" "s3" {
+  name = "${local.lambda_full_name}-s3"
+  description = join("", [
+    "Permissions for the ${local.lambda_full_name} Lambda to list and get objects in the ",
+    "${data.aws_s3_bucket.etl.id} S3 bucket in the Done/ and Incoming/ folders"
+  ])
+
+  policy = <<-EOF
+{
+  "Version": "2012-10-17",
```
[the remark this answers is not in the document]
Comment (mjburling) [reply ops/terraform/services/pipeline/modules/bfd_pipeline_slis/iam.tf:41]: > Versions are dates?

[Yes.](https://github.com/CMSgov/beneficiary-fhir-data/pull/1594#discussion_r1110044337)

File: ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+
+    BENEFICIARY = "bene"
+    CARRIER = "carrier"
+    DME = "dme"
+    HHA = "hha"
+    HOSPICE = "hospice"
+    INPATIENT = "inpatient"
+    OUTPATIENT = "outpatient"
+    PDE = "pde"
+    SNF = "snf"
```
[the remark this answers is not in the document]
Comment (mjburling) [reply ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py:69]: I don't think so. As far as the inputs to the bfd-pipeline ETL and CCW are concerned, they send us one beneficiary rif and eight claims rifs. These are the ones I'm familiar with. Does synthea have a 10th input for beneficiary history?

[the remark this answers is not in the document]
Comment (lsmitchell) [reply ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py:69]: Yes, Synthea sends us a beneficiary history file that has its own type.
image

[the remark this answers is not in the document]
Comment (malessi) [reply ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py:70]: I've updated the `Enum` to include `BENEFICIARY_HISTORY` as a file type, just in-case. The file type will only ever match if we receive a file with the name `beneficiary_history`, so it should never appear with CCW loads.

**Review by lsmitchell — APPROVED** [review]

**Review by mjburling — APPROVED** [review]

**Review by mjburling — APPROVED** [review]

**Review by lsmitchell — APPROVED** [review]

**Pull request merged**